File: sdk/src/agentforge/builder.py
import json
from typing import Any, Dict, List, Optional
from pydantic import BaseModel

from agentforge.types import (
    AgentDefinition,
    GraphDefinition,
    NodeConfig,
    EdgeConfig,
    SkillSpec,
    PolicyConfig,
    AgentModelConfig,
)

class AgentPolicy:
    def __init__(self) -> None:
        self._policy = PolicyConfig()

    def allow_tools(self, *tools: str) -> "AgentPolicy":
        if self._policy.allowed_tools is None:
            self._policy.allowed_tools = []
        self._policy.allowed_tools.extend(tools)
        return self

    def deny_tool(self, *tools: str) -> "AgentPolicy":
        self._policy.denied_tools.extend(tools)
        return self

    def require_approval_for(self, *tools: str) -> "AgentPolicy":
        self._policy.require_human_approval_for.extend(tools)
        return self

    def deny_input_pattern(self, pattern: str) -> "AgentPolicy":
        self._policy.deny_patterns.append(pattern)
        return self

    def max_cost(self, cost: float, currency: str = "USD") -> "AgentPolicy":
        if currency.upper() != "USD":
            pass
        self._policy.max_cost_usd = cost
        return self

    def max_steps(self, steps: int) -> "AgentPolicy":
        self._policy.max_graph_steps = steps
        return self

    def allow_fetch_only(self, *urls: str) -> "AgentPolicy":
        if self._policy.allowed_fetch_url_prefixes is None:
            self._policy.allowed_fetch_url_prefixes = []
        self._policy.allowed_fetch_url_prefixes.extend(urls)
        return self

    def max_message_history(self, n: int) -> "AgentPolicy":
        self._policy.max_message_history = n
        return self

    def context_compression_threshold(self, tokens: int) -> "AgentPolicy":
        self._policy.context_compression_threshold = tokens
        return self

    def build(self) -> PolicyConfig:
        return self._policy


class AgentBuilder:
    def __init__(self, name: str = "My Agent"):
        self._name = name
        self._description = None
        self._nodes: List[NodeConfig] = []
        self._edges: List[EdgeConfig] = []
        self._entry_point = None
        self._skills: List[SkillSpec] = []
        self._policy: Optional[PolicyConfig] = None
        self._model_config = AgentModelConfig()
        self._parallel_nodes: List[str] = []

    def description(self, desc: str) -> "AgentBuilder":
        self._description = desc
        return self

    def model(
        self,
        provider: str,
        model: str,
        temperature: float = 0.7,
        base_url: Optional[str] = None,
        options: Optional[Dict[str, Any]] = None,
    ) -> "AgentBuilder":
        self._model_config = AgentModelConfig(
            provider=provider,
            model=model,
            temperature=temperature,
            base_url=base_url,
            options=options or {},
        )
        return self

    def llm_node(self, id: str, system_prompt: str = "") -> "AgentBuilder":
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(
            id=id,
            type="llm",
            config={"system_prompt": system_prompt}
        ))
        return self

    def tool_node(self, id: str, tool_name: str) -> "AgentBuilder":
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(
            id=id,
            type="tool",
            config={"tool_name": tool_name}
        ))
        return self

    def asr_node(
        self,
        id: str,
        provider: str = "openai_whisper",
        language: str | None = None,
        filename: str | None = None,
    ) -> "AgentBuilder":
        config: Dict[str, Any] = {"provider": provider}
        if language:
            config["language"] = language
        if filename:
            config["filename"] = filename
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(id=id, type="asr", config=config))
        return self

    def tts_node(
        self,
        id: str,
        provider: str = "openai_tts",
        voice: str = "nova",
    ) -> "AgentBuilder":
        config: Dict[str, Any] = {"provider": provider, "voice": voice}
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(id=id, type="tts", config=config))
        return self

    def subagent_node(self, id: str, agent_id: str) -> "AgentBuilder":
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(
            id=id,
            type="subagent",
            config={"agent_id": agent_id}
        ))
        return self

    def custom_node(self, id: str, node_type: str, config: Dict[str, Any]) -> "AgentBuilder":
        if not self._entry_point:
            self._entry_point = id
        self._nodes.append(NodeConfig(
            id=id,
            type=node_type,
            config=config
        ))
        return self

    def edge(self, from_: str, to: str, condition: Optional[str] = None, condition_type: str = "always") -> "AgentBuilder":
        self._edges.append(EdgeConfig(
            from_=from_,
            to=to,
            condition=condition,
            condition_type=condition_type
        ))
        return self

    def parallel_nodes(self, *node_ids: str) -> "AgentBuilder":
        self._parallel_nodes.extend(node_ids)
        return self

    def policy(self, policy: AgentPolicy | PolicyConfig) -> "AgentBuilder":
        if isinstance(policy, AgentPolicy):
            self._policy = policy.build()
        else:
            self._policy = policy
        return self

    def skill(self, name: str, skill_type: str = "instruction", source_code: str = "", instructions: str = "") -> "AgentBuilder":
        self._skills.append(SkillSpec(
            name=name,
            skill_type=skill_type,
            source_code=source_code,
            instructions=instructions
        ))
        return self

    def build(self) -> AgentDefinition:
        return AgentDefinition(
            name=self._name,
            description=self._description,
            graph_definition=GraphDefinition(
                nodes=self._nodes,
                edges=self._edges,
                entry_point=self._entry_point,
                parallel_nodes=list(self._parallel_nodes),
            ),
            model_config=self._model_config,
            skills=self._skills,
            execution_policy=self._policy
        )

    def export_json(self, filepath: str) -> None:
        agent_def = self.build()
        with open(filepath, "w", encoding="utf-8") as f:
            f.write(agent_def.model_dump_json(indent=2, by_alias=True))

    def build_and_export(self, filepath: str) -> None:
        """Alias of export_json."""
        self.export_json(filepath)

    def run(self, inputs: Dict[str, Any]) -> Any:
        from agentforge.agent import LocalAgent
        agent_def = self.build()
        agent = LocalAgent(agent_def.model_dump(by_alias=True))
        return agent.invoke(inputs)

    async def arun(self, inputs: Dict[str, Any]) -> Any:
        from agentforge.agent import LocalAgent
        agent_def = self.build()
        agent = LocalAgent(agent_def.model_dump(by_alias=True))
        return await agent.ainvoke(inputs)

def Agent(name: str = "My Agent") -> AgentBuilder:
    return AgentBuilder(name)

File: sdk-js/src/builder.ts
import type {
  AgentDefinition,
  AgentModelConfig,
  ConditionType,
  EdgeConfig,
  GraphDefinition,
  NodeConfig,
  PolicyConfig,
  SkillOptions,
  SkillSpec,
} from "./types.js";
import { AgentClient, type AgentClientConfig } from "./client.js";

function cloneNode(node: NodeConfig): NodeConfig {
  return {
    ...node,
    config: { ...node.config },
  };
}

function cloneEdge(edge: EdgeConfig): EdgeConfig {
  return { ...edge };
}

function cloneSkill(skill: SkillSpec): SkillSpec {
  return {
    ...skill,
    ...(skill.metadata ? { metadata: { ...skill.metadata } } : {}),
  };
}

function clonePolicy(policy?: PolicyConfig): PolicyConfig | undefined {
  if (!policy) {
    return undefined;
  }

  return {
    ...(policy.allowed_tools ? { allowed_tools: [...policy.allowed_tools] } : {}),
    ...(policy.denied_tools?.length ? { denied_tools: [...policy.denied_tools] } : {}),
    ...(policy.allowed_fetch_url_prefixes
      ? { allowed_fetch_url_prefixes: [...policy.allowed_fetch_url_prefixes] }
      : {}),
    ...(policy.deny_patterns?.length ? { deny_patterns: [...policy.deny_patterns] } : {}),
    ...(policy.require_human_approval_for?.length
      ? { require_human_approval_for: [...policy.require_human_approval_for] }
      : {}),
    ...(policy.max_cost_usd !== undefined ? { max_cost_usd: policy.max_cost_usd } : {}),
    ...(policy.max_graph_steps !== undefined ? { max_graph_steps: policy.max_graph_steps } : {}),
    ...(policy.max_message_history !== undefined
      ? { max_message_history: policy.max_message_history }
      : {}),
    ...(policy.context_compression_threshold !== undefined
      ? { context_compression_threshold: policy.context_compression_threshold }
      : {}),
  };
}

function appendItems(values: string[] | undefined, items: string[]): string[] {
  return [...(values ?? []), ...items];
}

export class AgentPolicy {
  private policyConfig: PolicyConfig;

  constructor(initial?: PolicyConfig) {
    this.policyConfig = clonePolicy(initial) ?? {};
  }

  allowTools(...tools: string[]): this {
    this.policyConfig.allowed_tools = appendItems(this.policyConfig.allowed_tools, tools);
    return this;
  }

  denyTool(...tools: string[]): this {
    this.policyConfig.denied_tools = appendItems(this.policyConfig.denied_tools, tools);
    return this;
  }

  requireApprovalFor(...tools: string[]): this {
    this.policyConfig.require_human_approval_for = appendItems(
      this.policyConfig.require_human_approval_for,
      tools,
    );
    return this;
  }

  denyInputPattern(pattern: string): this {
    this.policyConfig.deny_patterns = appendItems(this.policyConfig.deny_patterns, [pattern]);
    return this;
  }

  maxCost(cost: number, currency = "USD"): this {
    void currency;
    this.policyConfig.max_cost_usd = cost;
    return this;
  }

  maxSteps(steps: number): this {
    this.policyConfig.max_graph_steps = steps;
    return this;
  }

  allowFetchOnly(...urls: string[]): this {
    this.policyConfig.allowed_fetch_url_prefixes = appendItems(
      this.policyConfig.allowed_fetch_url_prefixes,
      urls,
    );
    return this;
  }

  maxMessageHistory(n: number): this {
    this.policyConfig.max_message_history = n;
    return this;
  }

  contextCompressionThreshold(tokens: number): this {
    this.policyConfig.context_compression_threshold = tokens;
    return this;
  }

  build(): PolicyConfig {
    return clonePolicy(this.policyConfig) ?? {};
  }
}

export class AgentBuilder {
  private readonly agentName: string;
  private agentDescription: string | undefined;
  private nodes: NodeConfig[] = [];
  private edges: EdgeConfig[] = [];
  private entryPoint: string | undefined;
  private skills: SkillSpec[] = [];
  private executionPolicy: PolicyConfig | undefined;
  private parallelNodeIds: string[] = [];
  private modelConfig: AgentModelConfig = {
    provider: "openai",
    model: "gpt-4o",
    temperature: 0.7,
  };

  constructor(name = "My Agent") {
    this.agentName = name;
  }

  description(description: string): this {
    this.agentDescription = description;
    return this;
  }

  model(provider: string, model: string, temperature = 0.7): this {
    this.modelConfig = {
      provider,
      model,
      temperature,
    };
    return this;
  }

  llmNode(id: string, systemPrompt = ""): this {
    return this.addNode({
      id,
      type: "llm",
      config: { system_prompt: systemPrompt },
    });
  }

  toolNode(id: string, toolName: string): this {
    return this.addNode({
      id,
      type: "tool",
      config: { tool_name: toolName },
    });
  }

  subagentNode(id: string, agentId: string): this {
    return this.addNode({
      id,
      type: "subagent",
      config: { agent_id: agentId },
    });
  }

  customNode(id: string, nodeType: string, config: Record<string, unknown>): this {
    return this.addNode({
      id,
      type: nodeType,
      config,
    });
  }

  edge(from: string, to: string, condition?: string, conditionType: ConditionType = "always"): this {
    this.edges.push({
      from,
      to,
      ...(condition !== undefined ? { condition } : {}),
      condition_type: conditionType,
    });
    return this;
  }

  parallelNodes(...nodeIds: string[]): this {
    this.parallelNodeIds.push(...nodeIds);
    return this;
  }

  policy(policy: AgentPolicy | PolicyConfig): this {
    this.executionPolicy = policy instanceof AgentPolicy ? policy.build() : clonePolicy(policy);
    return this;
  }

  skill(name: string, options: SkillOptions = {}): this {
    const skill: SkillSpec = {
      name,
      skill_type: options.skillType ?? "instruction",
      ...(options.description ? { description: options.description } : {}),
      ...(options.sourceCode ? { source_code: options.sourceCode } : {}),
      ...(options.instructions ? { instructions: options.instructions } : {}),
      ...(options.metadata ? { metadata: options.metadata } : {}),
    };

    this.skills.push(skill);
    return this;
  }

  build(): AgentDefinition {
    const graphDefinition: GraphDefinition = {
      graph_schema_version: "1.0",
      nodes: this.nodes.map(cloneNode),
      edges: this.edges.map(cloneEdge),
      ...(this.entryPoint ? { entry_point: this.entryPoint } : {}),
      ...(this.parallelNodeIds.length ? { parallel_nodes: [...this.parallelNodeIds] } : {}),
    };

    const definition: AgentDefinition = {
      name: this.agentName,
      graph_definition: graphDefinition,
      model_config: { ...this.modelConfig },
      skills: this.skills.map(cloneSkill),
    };

    if (this.agentDescription) {
      definition.description = this.agentDescription;
    }

    if (this.executionPolicy) {
      definition.execution_policy = clonePolicy(this.executionPolicy) ?? {};
    }

    return definition;
  }

  toJSON(pretty = false): string {
    return JSON.stringify(this.build(), null, pretty ? 2 : undefined);
  }

  async push(apiUrl?: string, token?: string): Promise<{ id: string }> {
    const config: AgentClientConfig = {};
    if (apiUrl) config.apiUrl = apiUrl;
    if (token) config.token = token;
    const client = new AgentClient(config);
    return client.push(this.build());
  }

  private addNode(node: NodeConfig): this {
    if (!this.entryPoint) {
      this.entryPoint = node.id;
    }

    this.nodes.push(cloneNode(node));
    return this;
  }
}

export function Agent(name = "My Agent"): AgentBuilder {
  return new AgentBuilder(name);
}
